obs-ffmpeg: Fix reconnection for srt protocol and mpegts muxer

pkviet opened the pull request:
### Description
The mpegts output used with srt protocol has long been plagued by reconnection issues, leading to crashes.
This fixes (hopefully) the issue.

### Motivation and Context
The reconnection logic proved difficult to implement for several reasons:
- the reconnection with rtmp and the logic used in backend (in obs-output.c) is difficult to adapt to the mpegts muxer; it would require to track several bools, which make very easy to be sidetracked along the many paths of failure.
- on srt side, the srt_connect function can return a success in socket non-blocking mode and caller mode, which would trigger a signal to obs UI of OBS_OUTPUT_SUCCESS, while no connection has really been made.
- this could lead to crashes because the UI and the real state of the output might not be in sync, leading to crashes.

In order to fix the issue, we have:
- reworked the mpegts reconnect logic by moving the start and stop functions to a single thread controlled by a mutex, so that these two pipelines are now mutually exclusive. This avoids one major point of failure by preventing a call to start while stop has not finished cleaning up. We have followed the logic used in the whip output which is much simpler to implement than the 
multi-layered bools used in rtmp-stream.c.
- made explicit checks on the socket connection state on the srt protocol side, so that an OBS_OUTPUT_SUCCESS code is never returned before having ascertained that a successful connection was realized.

This fixes:
https://github.com/obsproject/obs-studio/issues/11062
https://github.com/obsproject/obs-studio/issues/11620
https://github.com/obsproject/obs-studio/issues/11721
https://github.com/obsproject/obs-studio/issues/12133
https://github.com/obsproject/obs-studio/issues/11885

### How Has This Been Tested?
Used the clumsy app to simulate a complete network failure.
OBS detects correctly a network disconnection.
It triggers correctly the reconnect thread from obs-output.c.
The reconnects start.
After a few reconnects, upon reestablishment of the network, obs manages to reconnect.

### Types of changes
- Bug fix (non-breaking change which fixes an issue)
- Code cleanup (non-breaking change which makes code smaller or more readable) 

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

stephematician commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -598,9 +607,9 @@ bool ffmpeg_mpegts_data_init(struct ffmpeg_output *stream, struct ffmpeg_data *d
 /* ------------------------------------------------------------------------- */
 
-static inline bool stopping(struct ffmpeg_output *output)
+static inline bool stopping(struct ffmpeg_output *stream)
```
I'm as guilty as anyone for doing this; but extensive renaming combined with many logic changes makes it harder to review the commit; would be great to separate them into two commits in future.

pkviet replied:
haha, that's fair. Followed your advice and split the renaming to its own commit; the renaming is because the logging function expects the struct to be named 'stream' but I was careless because i wrote the output as a mongrel between obs-ffmpeg-output and rtmp-stream.

ogmkp commented:
If I start a stream to srt://:port?mode=listener or srt://0.0.0.0:port?mode=listener or 127.0.0.1:port?mode=listener,
The connection (streaming) button still stuck and grayed, I have to kill OBS.

```
debug: obs_output_set_audio_encoder: Null 'output' parameter
debug: output 'adv_stream' (ffmpeg_mpegts_muxer) created
debug: [obs-nvenc: 'advanced_video_stream'] Loading up CUDA on device 0
info: [obs-nvenc: 'advanced_video_stream'] settings:
	codec:        H264
	rate_control: CBR
	bitrate:      1000
	keyint:       120
	preset:       p5
	tuning:       hq
	multipass:    qres
	profile:      high
	width:        1920
	height:       1080
	b-frames:     2
	b-ref-mode:   0
	lookahead:    false (0 frames)
	aq:           true

info: ---------------------------------
info: [FFmpeg aac encoder: 'Track1'] bitrate: 128, channels: 2, channel_layout: stereo, track: 1

info: [obs-ffmpeg mpegts muxer: 'adv_stream']: Output format name and long_name: mpegts, MPEG-TS (MPEG-2 Transport Stream)
info: [obs-ffmpeg mpegts muxer / libsrt]: libsrt version 1.5.4 loaded
```


If I stream to srt://ip:port?mode=caller with a destination port that is not open on other side, OBS crash with segfault.

```
debug: obs_output_set_audio_encoder: Null 'output' parameter
debug: output 'adv_stream' (ffmpeg_mpegts_muxer) created
debug: [obs-nvenc: 'advanced_video_stream'] Loading up CUDA on device 0
info: [obs-nvenc: 'advanced_video_stream'] settings:
	codec:        H264
	rate_control: CBR
	bitrate:      1000
	keyint:       120
	preset:       p5
	tuning:       hq
	multipass:    qres
	profile:      high
	width:        1920
	height:       1080
	b-frames:     2
	b-ref-mode:   0
	lookahead:    false (0 frames)
	aq:           true

info: ---------------------------------
info: [FFmpeg aac encoder: 'Track1'] bitrate: 128, channels: 2, channel_layout: stereo, track: 1

info: [obs-ffmpeg mpegts muxer: 'adv_stream']: Output format name and long_name: mpegts, MPEG-TS (MPEG-2 Transport Stream)
info: [obs-ffmpeg mpegts muxer / libsrt]: libsrt version 1.5.4 loaded
error: [obs-ffmpeg mpegts muxer: 'adv_stream']: Failed to open the url or invalid stream
error: [obs-ffmpeg mpegts muxer: 'adv_stream']: Failed to open the url
error: failed to start the mpegts output
Erreur de segmentation
```


Again, if I add pkt_size=1128 parameter in the srt url, OBS don't start the streaming and ask to look at log:


```
06:40:13.388754/obs*E:SRT.cc: LiveCC: payload size: 1316 exceeds maximum allowed 1128
error: [obs-ffmpeg mpegts muxer / libsrt]: Operation not supported: Incorrect use of Message API (sendmsg/recvmsg)
warning: process_packet: Error writing packet: Unknown error occurred
info: Output 'adv_stream': stopping
info: Output 'adv_stream': Total frames output: 3
info: Output 'adv_stream': Total drawn frames: 21
info: ---------------------------------
info: [obs-ffmpeg mpegts muxer / libsrt]: Session Summary
	time elapsed [0.3 sec]
	mean speed [0.0 Mbp]
	total bytes sent [0.0 MB]
	bytes retransmitted [0.0 %]
	bytes dropped [0.0 %]

info: ==== Streaming Stop ================================================
info: [obs-ffmpeg mpegts muxer / libsrt]: SRT connection closed
debug: Qavg: 65536.000

debug: 2 frames left in the queue on closing
```


When I open 2 srt in two media sources and when I re-apply the stream in one of them, both media source reset (like before the patch) but no reset on external player (ffplay) with the same url, its only in OBS.


```
06:43:58.002567/mp_media_thread*E:SRT.ea: EID:21 no sockets to check, this would deadlock
debug: Operation not supported: All sockets removed from epoll, waiting would deadlock

info: [Media Source 'Source média 2']: settings:
	input:                   srt://ip:port?streamid=read:test&latency=200000
	input_format:            
	speed:                   100
	is_looping:              no
	is_linear_alpha:         no
	is_hw_decoding:          yes
	is_clear_on_media_end:   no
	restart_on_activate:     no
	close_when_inactive:     no
	full_decode:             no
	ffmpeg_options:          
warning: MP: av_read_frame failed: Unknown error occurred (-1313558101)
warning: [Media Source 'Source média']: Disconnected. Reconnecting...
06:43:58.061258/mp_media_thread*E:SRT.ea: EID:23 no sockets to check, this would deadlock
debug: Operation not supported: All sockets removed from epoll, waiting would deadlock

debug: Connection to srt://ip:port?streamid=read:test&latency=200000 failed: Unknown error occurred

warning: MP: Failed to open media: 'srt://ip:port?streamid=read:test&latency=200000'
warning: [Media Source 'Source média 2']: Disconnected. Reconnecting...
debug: co located POCs unavailable

debug: co located POCs unavailable

debug: Timestamp for source 'Source média 2' jumped by '3875810308', expected value 85508535951, input value 89384346259
info: [Media Source 'Source média 2']: Reconnected.
debug: Timestamp for source 'Source média' jumped by '3819330843', expected value 85573533628, input value 89392864471
info: [Media Source 'Source média']: Reconnected.
```

pkviet commented:
> If I use the pkt_size parameter in the SRT output streaming url to reduce packet size (because of reduced MTU in the path), OBS can't connect and display and error in popup telling me to look at log. OBS 31.0.4 on linux
> 
are you reporting this for current release or specifically for this PR ? if there's the same issue on master, then post a specific issue; if it works on master, but is broken in the PR, please provide details about your test setup

pkviet commented:
@ogmkp i apologize for the harsh words but from a dev point of view, it's difficult when info is dumped and lacks enough context to make it actionable. Realize that the maintenance for srt is done in my free time on top of a regular job. I am not paid for it.
As stated already we can't possibly support all linux variations. I myself develop chiefly on windows so to optimize consideration it is better to provide steps on windows.
So please try to provide repro steps with the ppa binary or with the flatpak. It's to ensure that the issue does not stem from the reporter local setup but is generic enough to deserve our attention.
For your crash with pkt_size on master, create an issue for it after reproducing it with the ppa.
For the listener in a connecting state, it is because the connection and reconnection pipeline is tailored to rtmp. It would require changes in libobs obs-output.c that i am not prepared to do, such as creating a waiting state for connections. This is less trivial than it sounds. So right now it works as I intended it to. The alternative for me would be to just disable the listener if people are unhappy with it. When i test a listener, i connect it to a caller. Afterwards, the stream can be stopped. This was already discussed so i won't delve into more details.

ogmkp commented:
No problem, most of us are volunteers and we do it with passion. Whether we're a developer or a power user.
I don't build with PPA directions, because it have to be compliant with Ubuntu and flatpak is too sealed to be agile, and then I build on Debian...
Ok for listener, maybe a popup before going any further would be useful, anyway, it's to avoid any other issues.
It's your PR, you've got the upper hand on it, since all I do is only report problems, as long as it don't create new ones.

pkviet commented:
@ogmkp i could definitively add a log entry about the listener waiting for a connection. For a pop up that will have to be negotiated with our UI/UX people but yeah that's a good suggestion.
The plan forward is to make a real UI for SRT settings and to have the stats graphs. But there's interference with other plans for the other outputs so it is very slow.

pkviet commented:
@ogmkp managed to repro your pkt_size issue and fixed it.
Fixed also the crash for an invalid URL.
And added a log entry in listener mode. The popup requires much more work.

ogmkp commented:
> @ogmkp managed to repro your pkt_size issue and fixed it. Fixed also the crash for an invalid URL. And added a log entry in listener mode. The popup requires much more work.

Super great, I'll get back to you.

ogmkp commented:
Sorry, this I taking time longer than expected. I'll get back to you as soon as possible.

pkviet commented:
> What would be concrete steps to test the changed code?
> 
> As this PR changes a lot, it will take me some time to review it thoroughly and also test it.

Here are test steps for the various bugs which are fixed:
A) reconnection:
1. establish an srt connection to an srt server (Settings>Stream : srt://serverIP:PORT?mode=caller)  (or use FFmpeg as listener on another machine in the same LAN, through: 
`ffmpeg -re -i srt://IP:some_port?mode=listener -f mpegts udp://IP:other_port?pkt_size=1316` ).
Point vlc or another player to the udp to check the stream is valid.
2. Either physically disconnect your ethernet or use an app to simulate such a disconnect. On windows, i use 'clumsy' (https://jagt.github.io/clumsy/ ) ; on macOS, the Network Link Conditioner tool (installed now with XCode)  can be used. (If testing against FFmpeg as listener, you'll have to restart the command since FFmpeg will no longer listen once the connection is cut).
3. OBS will display reconnection attempts in the bottom status bar (make sure they're enabled in Settings > Advanced).
4. Re-enable the network.
5. Observe that the SRT stream gets reconnected to the server.

B) Invalid URL:
1. In Settings>Stream, modify a valid SRT url to an invalid one (change the port for instance or the IP).
2. Check that OBS no longer crashes but stops streaming with an error message.

C) Non default packet size:
1. In Settings>Stream, modify a valid SRT url : srt://serverIP:port?mode=caller&pkt_size=1128.
2. Verify that a server or ffmpeg listener accepts the stream and that it can be played through for instance: 
`ffmpeg -re -i srt://IP:some_port?mode=listener -f mpegts udp://IP:other_port?pkt_size=1128`

PatTheMav reviewed (commented):
On top of my other comments it seems that if an error happens late enough in the stream setup that the already created encoders are not properly torn down again, or is this a "common" OBS error?

```
warning: encoder 'advanced_video_stream': Cannot apply a new video_t object after the encoder has been initialized
warning: encoder 'advanced_video_stream': Cannot set the scaled resolution after the encoder has been initialized
warning: encoder 'advanced_video_stream': Cannot enable GPU scaling after the encoder has been initialized
```

Also it is possible to deadlock OBS by doing the following:

* Set a breakpoint in `libsrt_open` on the _successful_ return statement `return OBS_OUTPUT_SUCCESS`
* Start an FFmpeg SRT listener
* Start Streaming
* Leave the program execution suspended
* Watch FFmpeg terminating with the following error:
```
[in#0 @ 0x600002218a00] Error opening input: Input/output error
Error opening input file srt://127.0.0.1:12345?mode=listener.
Error opening input files: Input/output error
```
* Wait for another few seconds before continuing program execution
* Attempt to stop the stream

I've commented on the specific line of code that deadlocks in those situations.

Interestingly the error messages in the log warn about this, but probably for other reasons:
```
16:50:07.005609/*E:SRT.ea: EID:2 no sockets to check, this would deadlock
error: [obs-ffmpeg mpegts muxer / libsrt]: Operation not supported: All sockets removed from epoll, waiting would deadlock
```

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -772,259 +816,366 @@ static bool get_extradata(struct ffmpeg_output *stream)
-	for (size_t i = 0; i < output->packets.num; i++)
-		av_packet_free(output->packets.array + i);
-	da_free(output->packets);
+	if (os_atomic_load_bool(&stream->start_stop_thread_active))
+		pthread_join(stream->start_stop_thread, NULL);
```
pushed a temptative fix, where i move the start_stop_thread joining outside a mutex.
It seems to bring no adverse consequence except preventing a deadlock for me on macos

PatTheMav replied:
Those are probably timing-sensitive as well, if you hold up the right thread for long enough, other threads seem to plough ahead and shared state gets mixed up.

Probably somewhat mitigated by your recent change, so we can revisit those errors if they do occur during normal execution.

pkviet replied:
was a botched rebase.

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-srt.h:
```diff
@@ -735,7 +817,20 @@ static int libsrt_open(URLContext *h, const char *uri)
 #endif
 
-	return 0;
+	/* In caller mode, srt_connect won't fail even if there is no established connection in non-blocking mode;
+	 * so, we explicitly check for the socket status before okeying OBS_OUTPUT_SUCCESS .
```
done

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-srt.h:
```diff
@@ -205,6 +244,7 @@ static int libsrt_listen(int eid, SRTSOCKET fd, const struct sockaddr *addr, soc
 	/* Max streamid length plus an extra space for the terminating null character */
 	char streamid[513];
 	int streamid_len = sizeof(streamid);
+	blog(LOG_INFO, "[obs-ffmpeg mpegts muxer / libsrt]: listening for a connection ...");
```
done

RytoEX replied:
Change doesn't seem to be reflected here.

pkviet replied:
@_@ the rebase threw it out it seems, sorry; it should be done now

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -439,8 +452,15 @@ static inline int open_output_file(struct ffmpeg_output *stream, struct ffmpeg_d
-		if ((rist || srt) && (ret == OBS_OUTPUT_CONNECT_FAILED || ret == OBS_OUTPUT_INVALID_STREAM)) {
-			error("Failed to open the url or invalid stream");
+		if ((rist || srt)) {
+			if (ret == OBS_OUTPUT_CONNECT_FAILED || ret == OBS_OUTPUT_INVALID_STREAM)
+				error("Failed to open the url or invalid stream");
+			if (ret == OBS_OUTPUT_BAD_PATH)
+				error("URL is malformed");
+			if (ret == OBS_OUTPUT_DISCONNECTED)
+				error("Server unreachable");
+			if (ret == OBS_OUTPUT_ERROR)
+				error("I/O error");
```
agreed, replaced with a switch

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -772,259 +815,366 @@ static bool get_extradata(struct ffmpeg_output *stream)
+	/* a) get audio encoders & retrieve number of tracks */
 	obs_encoder_t *aencoders[MAX_AUDIO_MIXES];
 	int num_tracks = 0;
-
 	for (;;) {
 		obs_encoder_t *aencoder = obs_output_get_audio_encoder(stream->output, num_tracks);
 		if (!aencoder)
 			break;
-
```
done

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -772,259 +815,366 @@ static bool get_extradata(struct ffmpeg_output *stream)
-		ffmpeg_mpegts_full_stop(output);
-	} else {
-		obs_output_signal_stop(output->output, OBS_OUTPUT_SUCCESS);
+	/* Based on whip-output.cpp reconnect logic, coded by tt2468.
+	 * "signal" exists because we have to preserve the "running" state
+	 * across reconnect attempts. If we don't emit a signal if
+	 * something calls obs_output_stop() and it's reconnecting, you'll
+	 * desync the UI, as the output will be "stopped" and not
+	 * "reconnecting", but the "stop" signal will have never been
+	 * emitted.
+	 * We only clear bool 'running' if this is a user-requested stop,
+	 * not a reconnect-triggered stop
+	 */
```
fixed

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -365,7 +377,8 @@ static inline int allocate_custom_aviocontext(struct ffmpeg_output *stream, bool
 	buffer_size = UDP_DEFAULT_PAYLOAD_SIZE;
-
+	if (h->max_packet_size)
+		buffer_size = h->max_packet_size;
```
done

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -638,59 +661,75 @@ static void *ffmpeg_mpegts_create(obs_data_t *settings, obs_output_t *output)
+		/* Now wait for that stop to finish */
+		pthread_mutex_lock(&stream->start_stop_mutex);
+		if (os_atomic_load_bool(&stream->start_stop_thread_active))
+			pthread_join(stream->start_stop_thread, NULL);
```
done

pkviet commented:
**On hold.**
I have a regression between current iteration of the PR and earlier ones.

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -283,19 +283,23 @@ static inline int connect_mpegts_url(struct ffmpeg_output *stream, bool is_rist)
 	uc->url = (char *)url;
 	uc->max_packet_size = is_rist ? RIST_MAX_PAYLOAD_SIZE : SRT_LIVE_DEFAULT_PAYLOAD_SIZE;
+	if (stream->ff_data.config.srt_pkt_size)
+		uc->max_packet_size = stream->ff_data.config.srt_pkt_size;
```
fixed

pkviet commented:
**Update**
Comments addressed and regression fixed.

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -728,32 +795,36 @@ static int mpegts_process_packet(struct ffmpeg_output *output)
-
-			obs_output_signal_stop(output->output, code);
-			ffmpeg_mpegts_deactivate(output);
+			//pthread_detach(stream->write_thread);
```
my bad, i've been sloppy; fixed

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -728,32 +795,36 @@ static int mpegts_process_packet(struct ffmpeg_output *output)
-
-	os_atomic_set_bool(&output->active, false);
+	os_atomic_set_bool(&stream->stopping, true);
+	//stream->write_thread_active = false;
```
fixed

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -772,259 +843,385 @@ static bool get_extradata(struct ffmpeg_output *stream)
+static bool setup_muxer_settings(struct ffmpeg_output *stream, struct ffmpeg_cfg *config)
+{
+	/* Muxer & protocol settings */
+	/* This requires some UI to be written for the output.
+	 * at the service level unless one can load the output in the settings/stream screen.
```
chenged to: /* TODO: This will require some UI to be written for the mpegts output. */

pkviet commented on plugins/obs-ffmpeg/obs-ffmpeg-mpegts.c:
```diff
@@ -772,259 +843,385 @@ static bool get_extradata(struct ffmpeg_output *stream)
+	bool have_to_join = false;
+	pthread_t to_join;
+
+	/* macos fix: handover of pthread */
```
elbaorated into: /* macOS fix: handover of pthread; for some reason, maybe related to srt sockets, if the pthread joining is
 * done within the mutex, there can be a stall. So we just copy it, update the start_stop_thread_active volatile bool and join out of the mutex.*/

pkviet commented:
> Commit message nits:
> 
> * srt -> SRT
> * whip -> WHIP
> 
> Nits.

Fixed
